diff --git a/dep/_msquic/submodules/CMakeLists.txt b/dep/_msquic/submodules/CMakeLists.txt
--- a/dep/_msquic/submodules/CMakeLists.txt	
+++ b/dep/_msquic/submodules/CMakeLists.txt	
@@ -241,6 +241,8 @@
                     set(OPENSSL_CONFIG_CMD ${CMAKE_CURRENT_SOURCE_DIR}/${QUIC_OPENSSL}/Configure linux-armv4)
                 endif()
                 list(APPEND OPENSSL_CONFIG_FLAGS -latomic)
+            elseif (CMAKE_TARGET_ARCHITECTURE STREQUAL x64)
+                set(OPENSSL_CONFIG_CMD ${CMAKE_CURRENT_SOURCE_DIR}/${QUIC_OPENSSL}/Configure linux-x86_64)
             else()
                 set(OPENSSL_CONFIG_CMD ${CMAKE_CURRENT_SOURCE_DIR}/${QUIC_OPENSSL}/config
                             CC=${CMAKE_C_COMPILER} CXX=${CMAKE_CXX_COMPILER})
